sw/device/sca/sha3_serial.c

```diff
@@ -0,0 +1,182 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/dif/dif_kmac.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/sca/lib/prng.h"
+#include "sw/device/sca/lib/sca.h"
+#include "sw/device/sca/lib/simple_serial.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "kmac_regs.h"
+
+/**
+ * OpenTitan program for side-channel analysis of the absorb step of a KMAC128
+ * operation using a 128-bit key.
+ *
+ * This program implements the following simple serial commands:
+ *   - Set key ('k')*,
+ *   - Absorb ('p')*,
+ *   - Version ('v')+,
+ *   - Seed PRNG ('s')+,
+ * Commands marked with * are implemented in this file. Those marked with + are
+ * implemented in the simple serial library. See
+ * https://wiki.newae.com/SimpleSerial for details on the protocol.
+ */
+
+enum {
+  /**
+   * Key length in bytes.
+   */
+  kKeyLength = 16,
+  /**
+   * Message length in bytes.
+   */
+  kMessageLength = 16,
+};
+
+/**
+ * A handle to KMAC.
+ */
+static dif_kmac_t kmac;
+
+/**
+ * KMAC key.
+ *
+ * Used for caching the key in the 'k' (set key) command packet until it is used
+ * when handling a 'p' (absorb) command.
+ */
+static dif_kmac_key_t kmac_key;
+
+/**
+ * Blocks until KMAC is idle.
+ */
+static void kmac_block_until_idle(void) {
+  uint32_t reg;
+  do {
+    reg = mmio_region_read32(kmac.params.base_addr, KMAC_STATUS_REG_OFFSET);
+  } while (!bitfield_bit32_read(reg, KMAC_STATUS_SHA3_IDLE_BIT));
+}
+
+/**
+ * Resets KMAC to idle state.
+ */
+static void kmac_reset(void) {
+  mmio_region_write32(
+      kmac.params.base_addr, KMAC_CMD_REG_OFFSET,
+      bitfield_field32_write(0, KMAC_CMD_CMD_FIELD, KMAC_CMD_CMD_VALUE_DONE));
+  kmac_block_until_idle();
+}
+
+/**
+ * Issues a process command to KMAC and waits until absorb step is complete.
+ */
+static void kmac_absorb_end(void) {
+  uint32_t reg = 0;
+  reg = bitfield_field32_write(reg, KMAC_CMD_CMD_FIELD,
```

Comment:
We should discuss with @eunchan to add a parameter to delay the start of the operation for testing purposes,  and a TODO reference here. Maybe this is sufficient to switch to the call and sleep approach.

Reply:
Captured and added links to your and @vogelpi's comments in #7841 and added a TODO here.